fastify.js

```diff
@@ -585,6 +590,16 @@ function build (options) {
   function addHook (name, fn) {
     throwIfAlreadyStarted('Cannot call "addHook" when fastify instance is already started!')
 
+    if (name === 'onSend' || name === 'preSerialization') {
+      if (fn.constructor.name === 'AsyncFunction' && fn.length === 4) {
```

Comment:
I was working on this yesterday and one question popped up: what if someone were to use a function that returns a promise?

```
fastify.addHook('onSend', (request, reply, next) => new Promise(/* ... */));
```

This concern popped up for me when reading threads such as [this](https://github.com/tc39/ecmascript-asyncawait/issues/78) where it's pointed out that there isn't really a 'good' way of detecting an async function.

Reply:
Detecting an async function isn't a critical part of Fastify's functionality. This is just a way to help prevent people from making mistakes. If you really want to prevent every possible way someone could make a mistake then you'd have to use a linter.